zk/legacy_executor_verifier/legacy_executor_verifier.go

```diff
@@ -264,26 +178,95 @@ func (v *LegacyExecutorVerifier) handleRequest(ctx context.Context, request *Ver
 
 	previousBlock, _ := rawdb.ReadBlockByNumber(tx, blocks[0]-1)
 
-	ok, err := execer.Verify(payload, request, previousBlock.Root())
+	// eager promise will do the work as soon as called in a goroutine, then we can retrieve the result later
+	promise := NewPromise[*VerifierResponse](func() (*VerifierResponse, error) {
+		p := payload
+		r := request
+		root := previousBlock.Root()
+		failureResponse := &VerifierResponse{
+			BatchNumber: request.BatchNumber,
+			Valid:       false,
+			Witness:     witness,
+		}
+
+		ok, err2 := execer.Verify(p, r, root)
+		if err2 != nil {
+			return failureResponse, err2
+		}
+
+		response := &VerifierResponse{
+			BatchNumber: request.BatchNumber,
+			Valid:       ok,
+			Witness:     witness,
+		}
+
+		return response, nil
+	})
+
+	// add batch to the list of batches we've added
+	v.batchesMutex.Lock()
+	v.addedBatches[request.BatchNumber] = struct{}{}
+	v.batchesMutex.Unlock()
+
+	// add the promise to the list of promises
+	v.promises = append(v.promises, promise)
+	return promise, nil
+}
+
+func writeBatchToStream(result *VerifierResponse, hdb *hermez_db.HermezDbReader, roTx kv.Tx, v *LegacyExecutorVerifier) error {
+	blks, err := hdb.GetL2BlockNosByBatch(result.BatchNumber)
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	if ok {
-		// update the datastream now that we know the batch is OK
-		if err = server.WriteBlocksToStream(tx, hermezDb, v.streamServer, v.stream, blocks[0], blocks[len(blocks)-1], "verifier"); err != nil {
-			return true, err
+	if err := server.WriteBlocksToStream(roTx, hdb, v.streamServer, v.stream, blks[0], blks[len(blks)-1], "verifier"); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (v *LegacyExecutorVerifier) ConsumeResults(tx kv.RwTx) ([]*VerifierResponse, error) {
```

Comment:
We have to delete some entries from addedBatches in this function, because otherwise it will always grow in side. It would be best to do in this function. It is also invoked from the same thread that invokes "AddRequest" and "IsRequestAdded" so we still do not need mutex